{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5e0ac76c-6f30-4b78-9220-3cf8440d1a9b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import qiime2 as q2\n",
    "from qiime2 import Visualization\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "57f5161e-179d-492a-89b7-9dbaa7db0d32",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data = 'Data'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "79dce0c3-68e1-4d34-af67-133929a07e9b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_classified = 'Data/classified'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f391c8d3-913b-4d2f-9ee0-3f5fa8cba694",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_raw = 'Data/raw'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e1de3968-ed15-4a7c-88f1-329eecbf302f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_alpha = 'Data/diversity'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "6eb035b6-b692-4e20-9ea9-73a2b7cb41a7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.10/site-packages/unifrac/__init__.py:9: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n",
      "\u001b[32mSaved Visualization to: Data/diversity/alpha-rarefaction.qzv\u001b[0m\n",
      "\u001b[0m\u001b[?25h"
     ]
    }
   ],
   "source": [
    "! qiime diversity alpha-rarefaction \\\n",
    "    --i-table $data_classified/table-filtered.qza \\\n",
    "    --p-max-depth 10000 \\\n",
    "    --p-iterations 10 \\\n",
    "    --m-metadata-file $data_raw/20250913_metadata_ITS.tsv \\\n",
    "    --o-visualization $data_alpha/alpha-rarefaction.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "cda879fe-a4d4-4195-ba5c-92c771f34b68",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter server extension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/jovyan/data/1ec9a82b-7309-4564-9f43-c6e9e6d7ec18')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 1ec9a82b-7309-4564-9f43-c6e9e6d7ec18>"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load(f\"{data_alpha}/alpha-rarefaction.qzv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7768b0cb-0f09-4fbb-b8eb-ffd8abb868aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Bis hier ist alles gut!!!\n",
    "\n",
    "#Versuch die Metadaten erst nachher zu mergen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "70886434-c2b7-40eb-9b54-66d7accf9600",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "os.makedirs(\"Data/diversity/kmerizer_results\", exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "92bf2062-b9b3-4c2f-b1f2-3d1acfe6c4af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.10/site-packages/unifrac/__init__.py:9: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n"
     ]
    }
   ],
   "source": [
    "! qiime kmerizer seqs-to-kmers \\\n",
    "  --i-sequences Data/classified/rep-seqs-filtered.qza \\\n",
    "  --i-table Data/diversity/core-metrics-results/rarefied_table.qza \\\n",
    "  --p-kmer-size 100 \\\n",
    "  --p-no-tfidf True \\\n",
    "  --o-kmer-table Data/diversity/kmerizer_results/kmer-table.qza "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7420e8e5-5514-41f5-ba0f-4f7896e0797a",
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime diversity alpha \\\n",
    "  --i-table $data_diversity/kmerizer_results/kmer_table.qza \\\n",
    "  --p-metric pielou_e \\\n",
    "  --o-alpha-diversity $data_diversity/core-metrics-results/pielou_e.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c6b9433-0e6d-4465-ae00-ef0861f0b455",
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime diversity alpha \\\n",
    "  --i-table $data_diversity/kmerizer_results/kmer_table.qza \\\n",
    "  --p-metric observed_features \\\n",
    "  --o-alpha-diversity $data_diversity/kmerizer_results/observed_features.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6472dbd-dfb6-49ad-8b2b-c993f56159de",
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime diversity alpha \\\n",
    "  --i-table $data_diversity/kmerizer_results/kmer_table.qza \\\n",
    "  --p-metric shannon \\\n",
    "  --o-alpha-diversity $data_diversity/kmerizer_results/shannon.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "c42a6919-f2b0-4248-8fee-932e34380354",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "#######################################################################################\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#Ab hier ist wie in Ãœbung\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4d217ae8-e261-4794-a7a5-788217cab01d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.10/site-packages/unifrac/__init__.py:9: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n",
      "\u001b[32mSaved FeatureTable[Frequency] to: Data/diversity/core-metrics-results/rarefied_table.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[AlphaDiversity] to: Data/diversity/core-metrics-results/observed_features_vector.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[AlphaDiversity] to: Data/diversity/core-metrics-results/shannon_vector.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[AlphaDiversity] to: Data/diversity/core-metrics-results/evenness_vector.qza\u001b[0m\n",
      "\u001b[32mSaved DistanceMatrix to: Data/diversity/core-metrics-results/jaccard_distance_matrix.qza\u001b[0m\n",
      "\u001b[32mSaved DistanceMatrix to: Data/diversity/core-metrics-results/bray_curtis_distance_matrix.qza\u001b[0m\n",
      "\u001b[32mSaved PCoAResults to: Data/diversity/core-metrics-results/jaccard_pcoa_results.qza\u001b[0m\n",
      "\u001b[32mSaved PCoAResults to: Data/diversity/core-metrics-results/bray_curtis_pcoa_results.qza\u001b[0m\n",
      "\u001b[32mSaved Visualization to: Data/diversity/core-metrics-results/jaccard_emperor.qzv\u001b[0m\n",
      "\u001b[32mSaved Visualization to: Data/diversity/core-metrics-results/bray_curtis_emperor.qzv\u001b[0m\n",
      "\u001b[0m\u001b[?25h"
     ]
    }
   ],
   "source": [
    "! qiime diversity core-metrics \\\n",
    "  --i-table $data_classified/table-filtered.qza \\\n",
    "  --m-metadata-file $data_raw/20250913_metadata_ITS.tsv \\\n",
    "  --p-sampling-depth 1500 \\\n",
    "  --output-dir $data_alpha/core-metrics-results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ffdf141b-96ce-4601-8e4d-161e0344d378",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.10/site-packages/unifrac/__init__.py:9: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n",
      "Usage: \u001b[94mqiime kmerizer core-metrics\u001b[0m [OPTIONS]\n",
      "\n",
      "  Generate kmer counts from sequences and apply a collection of diversity\n",
      "  metrics (non-phylogenetic) to compare samples.\n",
      "\n",
      "\u001b[1mInputs\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--i-sequences\u001b[0m ARTIFACT \u001b[32mFeatureData[Sequence | RNASequence |\u001b[0m\n",
      "    \u001b[32mProteinSequence]\u001b[0m      Biological sequences to kmerize.          \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--i-table\u001b[0m ARTIFACT \u001b[32mFeatureTable[Frequency]\u001b[0m\n",
      "                          Frequencies of sequences per sample.      \u001b[35m[required]\u001b[0m\n",
      "\u001b[1mParameters\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--p-sampling-depth\u001b[0m INTEGER\n",
      "    \u001b[32mRange(1, None)\u001b[0m        The total frequency that each sample should be\n",
      "                          rarefied to prior to computing diversity metrics.\n",
      "                                                                    \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--m-metadata-file\u001b[0m METADATA...\n",
      "    (multiple arguments   The sample metadata to use in the emperor plots.\n",
      "     will be merged)                                                \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--p-kmer-size\u001b[0m INTEGER   Length of kmers to generate.           \u001b[35m[default: 16]\u001b[0m\n",
      "  \u001b[94m--p-tfidf\u001b[0m / \u001b[94m--p-no-tfidf\u001b[0m\n",
      "                          If True, kmers will be scored using TF-IDF and\n",
      "                          output frequencies will be weighted by scores. If\n",
      "                          False, kmers are counted without TF-IDF scores.\n",
      "                                                              \u001b[35m[default: False]\u001b[0m\n",
      "  \u001b[94m--p-max-df\u001b[0m VALUE \u001b[32mFloat % Range(0, 1, inclusive_end=True) | Int\u001b[0m\n",
      "                          Ignore kmers that have a frequency strictly higher\n",
      "                          than the given threshold. If float, the parameter\n",
      "                          represents a proportion of sequences, if an integer\n",
      "                          it represents an absolute count.      \u001b[35m[default: 1.0]\u001b[0m\n",
      "  \u001b[94m--p-min-df\u001b[0m VALUE \u001b[32mFloat % Range(0, 1) | Int\u001b[0m\n",
      "                          Ignore kmers that have a frequency strictly lower\n",
      "                          than the given threshold. If float, the parameter\n",
      "                          represents a proportion of sequences, if an integer\n",
      "                          it represents an absolute count.        \u001b[35m[default: 1]\u001b[0m\n",
      "  \u001b[94m--p-max-features\u001b[0m INTEGER\n",
      "                          If not None, build a vocabulary that only considers\n",
      "                          the top \u001b[4mmax-features\u001b[0m ordered by frequency (or TF-IDF\n",
      "                          score).                                   \u001b[35m[optional]\u001b[0m\n",
      "  \u001b[94m--p-with-replacement\u001b[0m / \u001b[94m--p-no-with-replacement\u001b[0m\n",
      "                          Rarefy with replacement by sampling from the\n",
      "                          multinomial distribution instead of rarefying\n",
      "                          without replacement.                \u001b[35m[default: False]\u001b[0m\n",
      "  \u001b[94m--p-n-jobs\u001b[0m VALUE \u001b[32mInt % Range(1, None) | Str % Choices('auto')\u001b[0m\n",
      "                          [beta methods only] - The number of concurrent jobs\n",
      "                          to use in performing this calculation. May not\n",
      "                          exceed the number of available physical cores. If\n",
      "                          \u001b[4mn-jobs\u001b[0m = 'auto', one job will be launched for each\n",
      "                          identified CPU core on the host.        \u001b[35m[default: 1]\u001b[0m\n",
      "  \u001b[94m--p-pc-dimensions\u001b[0m INTEGER\n",
      "                          Number of principal coordinate dimensions to keep\n",
      "                          for plotting.                           \u001b[35m[default: 3]\u001b[0m\n",
      "  \u001b[94m--p-color-by\u001b[0m TEXT       Categorical measure from the input Metadata that\n",
      "                          should be used for color-coding the scatterplot.\n",
      "                                                                    \u001b[35m[optional]\u001b[0m\n",
      "  \u001b[94m--p-norm\u001b[0m TEXT \u001b[32mChoices('None', 'l1', 'l2')\u001b[0m\n",
      "                          Normalization procedure applied to TF-IDF scores.\n",
      "                          Ignored if tfidf=False. l2: Sum of squares of vector\n",
      "                          elements is 1. l1: Sum of absolute values of vector\n",
      "                          elements is 1.                     \u001b[35m[default: 'None']\u001b[0m\n",
      "\u001b[1mOutputs\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--o-rarefied-table\u001b[0m ARTIFACT \u001b[32mFeatureTable[Frequency]\u001b[0m\n",
      "                          The resulting rarefied feature table.     \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--o-kmer-table\u001b[0m ARTIFACT \u001b[32mFeatureTable[Frequency]\u001b[0m\n",
      "                          Frequencies of kmers per sample.          \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--o-observed-features-vector\u001b[0m ARTIFACT \u001b[32mSampleData[AlphaDiversity]\u001b[0m\n",
      "                          Vector of Observed Kmers values by sample.\n",
      "                                                                    \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--o-shannon-vector\u001b[0m ARTIFACT \u001b[32mSampleData[AlphaDiversity]\u001b[0m\n",
      "                          Vector of Shannon diversity values by sample.\n",
      "                                                                    \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--o-jaccard-distance-matrix\u001b[0m ARTIFACT\n",
      "    \u001b[32mDistanceMatrix\u001b[0m        Matrix of Jaccard distances between pairs of\n",
      "                          samples.                                  \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--o-bray-curtis-distance-matrix\u001b[0m ARTIFACT\n",
      "    \u001b[32mDistanceMatrix\u001b[0m        Matrix of Bray-Curtis dissimilarities between pairs\n",
      "                          of samples.                               \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--o-jaccard-pcoa-results\u001b[0m ARTIFACT\n",
      "    \u001b[32mPCoAResults\u001b[0m           PCoA matrix computed from Jaccard distances between\n",
      "                          samples.                                  \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--o-bray-curtis-pcoa-results\u001b[0m ARTIFACT\n",
      "    \u001b[32mPCoAResults\u001b[0m           PCoA matrix computed from Bray-Curtis\n",
      "                          dissimilarities between samples.          \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--o-scatterplot\u001b[0m VISUALIZATION\n",
      "                          Scatterplot of results. Axes can be selected to\n",
      "                          display alpha diversity results or PCoA coordinates\n",
      "                          computed from Jaccard or Bray-Curtis.     \u001b[35m[required]\u001b[0m\n",
      "\u001b[1mMiscellaneous\u001b[0m:\n",
      "  \u001b[94m--output-dir\u001b[0m PATH       Output unspecified results to a directory\n",
      "  \u001b[94m--verbose\u001b[0m / \u001b[94m--quiet\u001b[0m     Display verbose output to stdout and/or stderr\n",
      "                          during execution of this action. Or silence output\n",
      "                          if execution is successful (silence is golden).\n",
      "  \u001b[94m--recycle-pool\u001b[0m TEXT     Use a cache pool for pipeline resumption. QIIME 2\n",
      "                          will cache your results in this pool for reuse by\n",
      "                          future invocations. These pool are retained until\n",
      "                          deleted by the user. If not provided, QIIME 2 will\n",
      "                          create a pool which is automatically reused by\n",
      "                          invocations of the same action and removed if the\n",
      "                          action is successful. Note: these pools are local to\n",
      "                          the cache you are using.\n",
      "  \u001b[94m--no-recycle\u001b[0m            Do not recycle results from a previous failed\n",
      "                          pipeline run or save the results from this run for\n",
      "                          future recycling.\n",
      "  \u001b[94m--parallel\u001b[0m              Execute your action in parallel. This flag will use\n",
      "                          your default parallel config.\n",
      "  \u001b[94m--parallel-config\u001b[0m FILE  Execute your action in parallel using a config at\n",
      "                          the indicated path.\n",
      "  \u001b[94m--example-data\u001b[0m PATH     Write example data and exit.\n",
      "  \u001b[94m--citations\u001b[0m             Show citations and exit.\n",
      "  \u001b[94m--use-cache\u001b[0m DIRECTORY   Specify the cache to be used for the intermediate\n",
      "                          work of this action. If not provided, the default\n",
      "                          cache under $TMP/qiime2/<uname> will be used.\n",
      "                          IMPORTANT FOR HPC USERS: If you are on an HPC system\n",
      "                          and are using parallel execution it is important to\n",
      "                          set this to a location that is globally accessible\n",
      "                          to all nodes in the cluster.\n",
      "  \u001b[94m--help\u001b[0m                  Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Missing option '--m-metadata-file'.\u001b[0m\n",
      "\u001b[0m\u001b[?25h"
     ]
    }
   ],
   "source": [
    "! qiime kmerizer core-metrics \\\n",
    "  --i-table $data_classified/table-filtered.qza \\\n",
    "  --i-sequences $data_classified/rep-seqs-filtered.qza \\\n",
    "  --p-sampling-depth 1500 \\\n",
    "  --p-kmer-size 8 \\\n",
    "  --m-metadata-file Data/raw/20250913_metadata_ITS.tsv \\\n",
    "  --output-dir $data_alpha/kmerizer-results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6d53eac0-590e-4eb0-bcbc-01bda61c4a33",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Metadata einlesen\n",
    "metadata_file = \"Data/raw/20250913_metadata_ITS.tsv\"\n",
    "df = pd.read_csv(metadata_file, sep='\\t')\n",
    "\n",
    "# Erste Spalte umbenennen auf '#SampleID'\n",
    "df.rename(columns={df.columns[0]: '#SampleID'}, inplace=True)\n",
    "\n",
    "# Als TSV wieder speichern (Tab-getrennt, ohne Index)\n",
    "df.to_csv(\"Data/raw/20250913_metadata_ITS_fixed.tsv\", sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a7598e86-b1e3-4618-abae-a793d228e002",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.10/site-packages/unifrac/__init__.py:9: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n",
      "Usage: \u001b[94mqiime kmerizer core-metrics\u001b[0m [OPTIONS]\n",
      "\n",
      "  Generate kmer counts from sequences and apply a collection of diversity\n",
      "  metrics (non-phylogenetic) to compare samples.\n",
      "\n",
      "\u001b[1mInputs\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--i-sequences\u001b[0m ARTIFACT \u001b[32mFeatureData[Sequence | RNASequence |\u001b[0m\n",
      "    \u001b[32mProteinSequence]\u001b[0m      Biological sequences to kmerize.          \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--i-table\u001b[0m ARTIFACT \u001b[32mFeatureTable[Frequency]\u001b[0m\n",
      "                          Frequencies of sequences per sample.      \u001b[35m[required]\u001b[0m\n",
      "\u001b[1mParameters\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--p-sampling-depth\u001b[0m INTEGER\n",
      "    \u001b[32mRange(1, None)\u001b[0m        The total frequency that each sample should be\n",
      "                          rarefied to prior to computing diversity metrics.\n",
      "                                                                    \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--m-metadata-file\u001b[0m METADATA...\n",
      "    (multiple arguments   The sample metadata to use in the emperor plots.\n",
      "     will be merged)                                                \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--p-kmer-size\u001b[0m INTEGER   Length of kmers to generate.           \u001b[35m[default: 16]\u001b[0m\n",
      "  \u001b[94m--p-tfidf\u001b[0m / \u001b[94m--p-no-tfidf\u001b[0m\n",
      "                          If True, kmers will be scored using TF-IDF and\n",
      "                          output frequencies will be weighted by scores. If\n",
      "                          False, kmers are counted without TF-IDF scores.\n",
      "                                                              \u001b[35m[default: False]\u001b[0m\n",
      "  \u001b[94m--p-max-df\u001b[0m VALUE \u001b[32mFloat % Range(0, 1, inclusive_end=True) | Int\u001b[0m\n",
      "                          Ignore kmers that have a frequency strictly higher\n",
      "                          than the given threshold. If float, the parameter\n",
      "                          represents a proportion of sequences, if an integer\n",
      "                          it represents an absolute count.      \u001b[35m[default: 1.0]\u001b[0m\n",
      "  \u001b[94m--p-min-df\u001b[0m VALUE \u001b[32mFloat % Range(0, 1) | Int\u001b[0m\n",
      "                          Ignore kmers that have a frequency strictly lower\n",
      "                          than the given threshold. If float, the parameter\n",
      "                          represents a proportion of sequences, if an integer\n",
      "                          it represents an absolute count.        \u001b[35m[default: 1]\u001b[0m\n",
      "  \u001b[94m--p-max-features\u001b[0m INTEGER\n",
      "                          If not None, build a vocabulary that only considers\n",
      "                          the top \u001b[4mmax-features\u001b[0m ordered by frequency (or TF-IDF\n",
      "                          score).                                   \u001b[35m[optional]\u001b[0m\n",
      "  \u001b[94m--p-with-replacement\u001b[0m / \u001b[94m--p-no-with-replacement\u001b[0m\n",
      "                          Rarefy with replacement by sampling from the\n",
      "                          multinomial distribution instead of rarefying\n",
      "                          without replacement.                \u001b[35m[default: False]\u001b[0m\n",
      "  \u001b[94m--p-n-jobs\u001b[0m VALUE \u001b[32mInt % Range(1, None) | Str % Choices('auto')\u001b[0m\n",
      "                          [beta methods only] - The number of concurrent jobs\n",
      "                          to use in performing this calculation. May not\n",
      "                          exceed the number of available physical cores. If\n",
      "                          \u001b[4mn-jobs\u001b[0m = 'auto', one job will be launched for each\n",
      "                          identified CPU core on the host.        \u001b[35m[default: 1]\u001b[0m\n",
      "  \u001b[94m--p-pc-dimensions\u001b[0m INTEGER\n",
      "                          Number of principal coordinate dimensions to keep\n",
      "                          for plotting.                           \u001b[35m[default: 3]\u001b[0m\n",
      "  \u001b[94m--p-color-by\u001b[0m TEXT       Categorical measure from the input Metadata that\n",
      "                          should be used for color-coding the scatterplot.\n",
      "                                                                    \u001b[35m[optional]\u001b[0m\n",
      "  \u001b[94m--p-norm\u001b[0m TEXT \u001b[32mChoices('None', 'l1', 'l2')\u001b[0m\n",
      "                          Normalization procedure applied to TF-IDF scores.\n",
      "                          Ignored if tfidf=False. l2: Sum of squares of vector\n",
      "                          elements is 1. l1: Sum of absolute values of vector\n",
      "                          elements is 1.                     \u001b[35m[default: 'None']\u001b[0m\n",
      "\u001b[1mOutputs\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--o-rarefied-table\u001b[0m ARTIFACT \u001b[32mFeatureTable[Frequency]\u001b[0m\n",
      "                          The resulting rarefied feature table.     \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--o-kmer-table\u001b[0m ARTIFACT \u001b[32mFeatureTable[Frequency]\u001b[0m\n",
      "                          Frequencies of kmers per sample.          \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--o-observed-features-vector\u001b[0m ARTIFACT \u001b[32mSampleData[AlphaDiversity]\u001b[0m\n",
      "                          Vector of Observed Kmers values by sample.\n",
      "                                                                    \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--o-shannon-vector\u001b[0m ARTIFACT \u001b[32mSampleData[AlphaDiversity]\u001b[0m\n",
      "                          Vector of Shannon diversity values by sample.\n",
      "                                                                    \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--o-jaccard-distance-matrix\u001b[0m ARTIFACT\n",
      "    \u001b[32mDistanceMatrix\u001b[0m        Matrix of Jaccard distances between pairs of\n",
      "                          samples.                                  \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--o-bray-curtis-distance-matrix\u001b[0m ARTIFACT\n",
      "    \u001b[32mDistanceMatrix\u001b[0m        Matrix of Bray-Curtis dissimilarities between pairs\n",
      "                          of samples.                               \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--o-jaccard-pcoa-results\u001b[0m ARTIFACT\n",
      "    \u001b[32mPCoAResults\u001b[0m           PCoA matrix computed from Jaccard distances between\n",
      "                          samples.                                  \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--o-bray-curtis-pcoa-results\u001b[0m ARTIFACT\n",
      "    \u001b[32mPCoAResults\u001b[0m           PCoA matrix computed from Bray-Curtis\n",
      "                          dissimilarities between samples.          \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--o-scatterplot\u001b[0m VISUALIZATION\n",
      "                          Scatterplot of results. Axes can be selected to\n",
      "                          display alpha diversity results or PCoA coordinates\n",
      "                          computed from Jaccard or Bray-Curtis.     \u001b[35m[required]\u001b[0m\n",
      "\u001b[1mMiscellaneous\u001b[0m:\n",
      "  \u001b[94m--output-dir\u001b[0m PATH       Output unspecified results to a directory\n",
      "  \u001b[94m--verbose\u001b[0m / \u001b[94m--quiet\u001b[0m     Display verbose output to stdout and/or stderr\n",
      "                          during execution of this action. Or silence output\n",
      "                          if execution is successful (silence is golden).\n",
      "  \u001b[94m--recycle-pool\u001b[0m TEXT     Use a cache pool for pipeline resumption. QIIME 2\n",
      "                          will cache your results in this pool for reuse by\n",
      "                          future invocations. These pool are retained until\n",
      "                          deleted by the user. If not provided, QIIME 2 will\n",
      "                          create a pool which is automatically reused by\n",
      "                          invocations of the same action and removed if the\n",
      "                          action is successful. Note: these pools are local to\n",
      "                          the cache you are using.\n",
      "  \u001b[94m--no-recycle\u001b[0m            Do not recycle results from a previous failed\n",
      "                          pipeline run or save the results from this run for\n",
      "                          future recycling.\n",
      "  \u001b[94m--parallel\u001b[0m              Execute your action in parallel. This flag will use\n",
      "                          your default parallel config.\n",
      "  \u001b[94m--parallel-config\u001b[0m FILE  Execute your action in parallel using a config at\n",
      "                          the indicated path.\n",
      "  \u001b[94m--example-data\u001b[0m PATH     Write example data and exit.\n",
      "  \u001b[94m--citations\u001b[0m             Show citations and exit.\n",
      "  \u001b[94m--use-cache\u001b[0m DIRECTORY   Specify the cache to be used for the intermediate\n",
      "                          work of this action. If not provided, the default\n",
      "                          cache under $TMP/qiime2/<uname> will be used.\n",
      "                          IMPORTANT FOR HPC USERS: If you are on an HPC system\n",
      "                          and are using parallel execution it is important to\n",
      "                          set this to a location that is globally accessible\n",
      "                          to all nodes in the cluster.\n",
      "  \u001b[94m--help\u001b[0m                  Show this message and exit.\n",
      "\n",
      "\u001b[33m                  There were some problems with the command:                  \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/12) Invalid value for '--i-table': /table-filtered.qza does not exist.\u001b[0m\n",
      "\u001b[31m\u001b[1m (2/12) Invalid value for '--i-sequences': /rep-seqs-filtered.qza does not\n",
      "  exist.\u001b[0m\n",
      "\u001b[31m\u001b[1m (3/12) Invalid value for '--output-dir': '/kmerizer-results/' is not a\n",
      "  writable directory, cannot write output to it.\u001b[0m\n",
      "\u001b[31m\u001b[1m (4/12) Missing option '--o-rarefied-table'. (\"--output-dir\" may also be used)\u001b[0m\n",
      "\u001b[31m\u001b[1m (5/12) Missing option '--o-kmer-table'. (\"--output-dir\" may also be used)\u001b[0m\n",
      "\u001b[31m\u001b[1m (6/12) Missing option '--o-observed-features-vector'. (\"--output-dir\" may\n",
      "  also be used)\u001b[0m\n",
      "\u001b[31m\u001b[1m (7/12) Missing option '--o-shannon-vector'. (\"--output-dir\" may also be used)\u001b[0m\n",
      "\u001b[31m\u001b[1m (8/12) Missing option '--o-jaccard-distance-matrix'. (\"--output-dir\" may also\n",
      "  be used)\u001b[0m\n",
      "\u001b[31m\u001b[1m (9/12) Missing option '--o-bray-curtis-distance-matrix'. (\"--output-dir\" may\n",
      "  also be used)\u001b[0m\n",
      "\u001b[31m\u001b[1m (10/12) Missing option '--o-jaccard-pcoa-results'. (\"--output-dir\" may also\n",
      "  be used)\u001b[0m\n",
      "\u001b[31m\u001b[1m (11/12) Missing option '--o-bray-curtis-pcoa-results'. (\"--output-dir\" may\n",
      "  also be used)\u001b[0m\n",
      "\u001b[31m\u001b[1m (12/12) Missing option '--o-scatterplot'. (\"--output-dir\" may also be used)\u001b[0m\n",
      "\u001b[0m\u001b[?25h"
     ]
    }
   ],
   "source": [
    "! qiime kmerizer core-metrics \\\n",
    "  --i-table $data_classified/table-filtered.qza \\\n",
    "  --i-sequences $data_classified/rep-seqs-filtered.qza \\\n",
    "  --m-metadata-file Data/raw/20250913_metadata_ITS_fixed.tsv \\\n",
    "  --p-sampling-depth 1500 \\\n",
    "  --p-kmer-size 8 \\\n",
    "  --output-dir $data_alpha/kmerizer-results \\\n",
    "  --verbose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "d96b5bd8-855d-4482-b7af-9c80b1e6f09f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ls: cannot access 'Data/diversity/kmerizer-results': No such file or directory\n"
     ]
    }
   ],
   "source": [
    "ls $data_alpha/kmerizer-results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecb57b39-dbf7-4779-9ea0-c0b63e384c07",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "QIIME 2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
